### ISSUE TITLE
insert_overwrite incremental strategy in combination with required partition_filter fails when there are zero rows to insert 

### ISSUE THREAD

**Issue by dhlstrm** [dbt-labs/dbt-bigquery#319]
### Is this a new bug in dbt-bigquery?

- [X] I believe this is a new bug in dbt-bigquery
- [X] I have searched the existing issues, and I could not find an existing issue for this bug

### Current Behavior

If a model has `insert_overwrite` incremental strategy and `require_partition_filter` set to `true`, an incremental run that tries to add zero (0) rows will fail and return ```Cannot query over table ***** without a filter over column(s) 'column_name' that can be used for partition elimination at [**:**]``` from BigQuery. The `_tmp` table then remains in the database since the error prevents it from being dropped.

### Expected Behavior

If no rows/partitions are found in the `_tmp` table, dbt should not proceed with trying to query the source table and skip to deleting the `_tmp` table. 

### Steps To Reproduce

1. Create an incremental model with `require_partition_filter` set to `true`, and `incremental_strategy` set to `insert_overwrite`, and any `partition_by` settings.
2. Run an incremental run on the model and make sure the query results in zero (0) rows.

### Relevant log output

```shell
The run query from "run" catalogue:

      -- generated script to merge partitions into `*****_daily`
      declare dbt_partitions_for_replacement array<date>;

      
      
        -- 1. temp table already exists, we used it to check for schema changes
      

      -- 2. define partitions to update
      set (dbt_partitions_for_replacement) = (
          select as struct
              array_agg(distinct session_date)
          from `*****_tmp`
      );

      
      -- 3. run the merge statement
      

    merge into `*****_daily` as DBT_INTERNAL_DEST
        using (
        select * from `*****_tmp`
      ) as DBT_INTERNAL_SOURCE
        on FALSE

    when not matched by source
         and DBT_INTERNAL_DEST.session_date in unnest(dbt_partitions_for_replacement) 
        then delete

    when not matched then insert
        (`col1`, `col2`)
    values
        (`col1`, `col2`)
;

      -- 4. clean up the temp table
      drop table if exists `*****_tmp`
```

### Environment

```markdown
- OS: MacOS Monterey
- Python: 3.8.9
- dbt-core: 1.2.0
- dbt-bigquery: 1.2.0
```

**Comment by callenilsson** [comment]
If anyone knows how to fix this bug, please help! I've scheduled dbt to run every hour, but since it fails when there are 0 new rows, it causes a LOT of problems for me.

One way to "solve" it is by temporarily disabling the `require_partition_filter` when dbt runs, and then enabling it after completed:
```
{{ config(
    ...
    pre_hook = "alter table if exists {{ this }} set options (require_partition_filter = false)",
    post_hook = "alter table if exists {{ this }} set options (require_partition_filter = true)"
) }}
```

But unfortunately that doesn't work for me because I've defined my tables in Terraform, and for some reason dbt cannot overwrite the `require_partition_filter` when the tables have been created externally. Nothing just happens. It only works if dbt initially created the tables from scratch (another bug?). Problem is I really need to define my tables in Terraform first to force some columns to be `REQUIRED` rather than `NULLABLE` in BigQuery, since dbt has no support for that yet (I think?).

The only solution I'm left with is to run the `dbt run` command as a Python subprocess, and wrap it in a try-except block to ignore the error (so it won't alert me every time it fails in the cloud). But this is so ugly that it makes me puke 🤮
```
import subprocess
res = subprocess.run("dbt run", shell=True, stdout=subprocess.PIPE, universal_newlines=True)
try:
    res.check_returncode()
except subprocess.CalledProcessError as e:
    if "without a filter over column(s)" not in res.stdout:
        raise e
```

**Comment by callenilsson** [comment]
I may have found what's causing this. If the destination table has the following config:
```
{{ config(
    materialized = "incremental",
    incremental_strategy = "insert_overwrite",
    require_partition_filter = true
    ...
) }}
```

And `dbt run ...` has no new rows to merge into the table, then BigQuery will throw this error since it's trying to perform `MERGE INTO ...` a table that strictly requires a partition filter. But when no partitions are selected, this error is raised.

The line I think is responsible for this is [this](https://github.com/dbt-labs/dbt-bigquery/blob/eee2790c324010570c310df20e7b0988a7ef1726/dbt/include/bigquery/macros/materializations/incremental.sql#L61), in `1.3.latest` branch, or [this](https://github.com/dbt-labs/dbt-bigquery/blob/2d4c92b94ad5ceb5b20688b51eb2033848d97592/dbt/include/bigquery/macros/materializations/incremental_strategy/insert_overwrite.sql#L108) in `main` branch. I think the `dbt_partitions_for_replacement` variable should have a default `cast(null as timestamp)` if empty.

Although when I tried this, it created a new empty `__NULL__` partition in BigQuery with 0 rows in it. I'm not sure if that could cause errors in other unexpected places. The same thing happens when using the pre/post_hook trick in my previous post, so I suppose that's just a manual way of achieving the same thing.

Anyway, I'm not sure how to fix this exactly in the dbt source code, so if anyone else want to help out here, please do!

**Comment by callenilsson** [comment]
I should mention that this happens not only for `incremental_strategy="insert_overwrite"`, but for `incremental_strategy="merge"` as well. This is a huge problem since new data haven't necessarily always been added to the source table on every run.

**Comment by gareginordyan** [comment]
This would be helpful for Fivetran as well!

**Comment by Fleid** [comment]
@callenilsson you got it.
Outside of the incremental logic (that should be taken care of in the model SQL statement), the culprits are:

- For `incremental_strategy="merge"`: in [`default__get_merge_sql`](https://github.com/dbt-labs/dbt-core/blob/f841a7ca76b44429eb94174bc5d0c2fecbf2763b/core/dbt/include/global_project/macros/materializations/models/incremental/merge.sql#L7), the predicate generation logic should take that case in consideration, it's used [line 37](https://github.com/dbt-labs/dbt-core/blob/f841a7ca76b44429eb94174bc5d0c2fecbf2763b/core/dbt/include/global_project/macros/materializations/models/incremental/merge.sql#L37)
- For `incremental_strategy="insert_overwrite"`: in ['default__get_insert_overwrite_merge_sql'](https://github.com/dbt-labs/dbt-core/blob/f841a7ca76b44429eb94174bc5d0c2fecbf2763b/core/dbt/include/global_project/macros/materializations/models/incremental/merge.sql#L107), same thing but this time on `when not matched by source`

I'm slowly realizing that we don't have a dispatch version of that one in BigQuery yet... But that would be where I add this bit of logic. We would also need to have the other merge macros checked for the same issue. @jtcohen6 I'd appreciate a sanity check on that plan ;)

At the moment I'll make this `help_wanted`, as I don't see us being able to tackle it in the short run.

**Comment by MatthieuBlais** [comment]
That's not solving the original issue, but as a workaround for those blocked by this (and using dbt 1.4+), it seems possible to use `copy_partitions: true` and skip the merge operation raising the error. https://docs.getdbt.com/blog/bigquery-ingestion-time-partitioning-and-partition-copy-with-dbt

```
{{ config(
    materialized = 'incremental',
    incremental_strategy = 'insert_overwrite',
    partition_by = {
      "field": "day",
      "data_type": "date",
      "copy_partitions": true
    }
) }}
```

**Comment by tnk-ysk** [comment]
I was trying to solve this issue, but it seems like this problem no longer occurs, probably due to a fix in bigquery.
